kubeca/certinit/certreq.go

```diff
@@ -0,0 +1,166 @@
+package certinit
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/ekspand/trusty/pkg/csr"
+	"github.com/ekspand/trusty/pkg/inmemcrypto"
+	"github.com/juju/errors"
+	capi "k8s.io/api/certificates/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	randAlphabet = "0123456789abcdefghijklmnopqrstuvwxyz"
+)
+
+var (
+	randAlphabetLength = big.NewInt(int64(len(randAlphabet)))
+)
+
+var usages = map[string][]capi.KeyUsage{
+	"peer": {
+		capi.UsageDigitalSignature,
+		capi.UsageKeyEncipherment,
+		capi.UsageServerAuth,
+		capi.UsageClientAuth,
+	},
+	"client": {
+		capi.UsageDigitalSignature,
+		capi.UsageKeyEncipherment,
+		capi.UsageClientAuth,
+	},
+	"server": {
+		capi.UsageDigitalSignature,
+		capi.UsageKeyEncipherment,
+		capi.UsageServerAuth,
+	},
+}
+
+func (r *Request) requestCertificate(ctx context.Context, client MinCertificates) error {
+	issuerAndProfile := strings.Split(r.SignerName, "/")
+	if len(issuerAndProfile) != 2 {
+		return errors.Errorf("unsupported signer: " + r.SignerName)
+	}
+
+	profileUsages := usages[issuerAndProfile[1]]
+	if profileUsages == nil {
+		return errors.Errorf("unsupported profile: " + r.SignerName)
+	}
+
+	prov := csr.NewProvider(inmemcrypto.NewProvider())
+	req := csr.CertificateRequest{
+		KeyRequest: prov.NewKeyRequest("", "ECDSA", 256, csr.SigningKey),
+		SAN:        r.san,
+	}
+
+	pemCsr, pemKeyBytes, _, _, err := prov.CreateRequestAndExportKey(&req)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	keyFile := path.Join(r.CertDir, "tls.key")
+	if err := ioutil.WriteFile(keyFile, pemKeyBytes, 0600); err != nil {
+		return errors.Annotate(err, "unable to safe key")
+	}
+
+	logger.Infof("wrote %s", keyFile)
+
+	csrFile := path.Join(r.CertDir, "tls.csr")
+	if err := ioutil.WriteFile(csrFile, pemCsr, 0644); err != nil {
+		return errors.Annotate(err, "unable to safe CSR")
+	}
+
+	logger.Infof("wrote %s", csrFile)
+
+	// Submit a certificate signing request, wait for it to be approved, then save
+	// the signed certificate to the file system.
+	certificateSigningRequestName := r.requestName()
+	certificateSigningRequest := &capi.CertificateSigningRequest{
+		TypeMeta: metaV1.TypeMeta{
+			Kind:       "CertificateSigningRequest",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metaV1.ObjectMeta{
+			Name:   certificateSigningRequestName,
+			Labels: r.labelsMap,
+		},
+		Spec: capi.CertificateSigningRequestSpec{
+			Request:    pemCsr,
+			Usages:     profileUsages,
+			SignerName: r.SignerName,
+			Extra: map[string]capi.ExtraValue{
+				"issuer":  []string{issuerAndProfile[0]},
+				"profile": []string{issuerAndProfile[1]},
+			},
+		},
+	}
+
+	_, err = client.Get(ctx, certificateSigningRequestName, metaV1.GetOptions{})
+	if err != nil {
+		_, err = client.Create(ctx, certificateSigningRequest, metaV1.CreateOptions{})
```

Comment:
The previous err is lost. At least log it as warning otherwise we never know why client.Get above failed.